### ISSUE TITLE
Color Scheme Simualation

### ISSUE THREAD

**Issue by janodvarko** [firefox-devtools/ux#129]
UX discussion for "Color Scheme Simulation" feature in DevTools

Summary:
Adds an option to simulate different color schemes allowing to test @prefers-color-scheme media queries. Using this media query lets your style sheet specify whether it prefers a light or dark user interface. This feature lets you test your code without having to change settings in your browser (or operating system, if the browser follows a system-wide color scheme setting).

Mini-PRD: https://docs.google.com/document/d/1M1YMw83k2wmW7FAnqnIC3DdkPElccxypsCLuWyoIZJY/edit#heading=h.ra7dtg5dt199

Original bug:
Bug 1550804 - Add color scheme simulation to the inspector
Introduced 4 states toggle button (null, no-preference, light, dark)

Meta: https://bugzilla.mozilla.org/show_bug.cgi?id=1137699

Hidden behind a pref:
devtools.inspector.color-scheme-simulation.enabled

MDN: https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Experimental_features#Color_scheme_simulation

The goals should be to polish/simplify the UI (e.g. no-preference doesn't exist anymore)

Others:
* Chrome has 3 states (seems to be clearer UI)
* Safari has 2 states (picking the opposite, default no-emulation)

@violasong @digitarald @martinbalfanz 

Honza

**Comment by janodvarko** [comment]
This feature might block progress on Fission see:

[Bug 1591120](https://bugzilla.mozilla.org/show_bug.cgi?id=1591120) - nsIContentViewer::emulatePrefersColorScheme should be forward to children remote iframes

**Comment by fvsch** [comment]
This was implemented with a button that cycles between 3 states (no-preference, forced-light, forced-dark), and there’s a bit of feedback on Twitter showing it can be puzzling. Here was my analysis: https://twitter.com/fvsch/status/1359911299409412096

I proposed an alternative design with two buttons, which can handle the 3 states:
- no active button = no preference
- dark or light button active = light or dark preference
- each button should disable the other when it becomes enabled

Figma design:
https://www.figma.com/file/9bQ5m4ETFQ3fs6dERCXs2V/Forced-Light-Dark-Mode-Buttons

Icon code, should be ready to use in mozilla-central:
https://gist.github.com/fvsch/1eec545f9fc0c42d4050634f009cfbbf

The icons are designed to be rendered at 16px, but if you need to make them smaller, both 15px and 14px seems to render correctly as well on 1x displays (since it’s a circular design no straight strokes, it doesn’t have as much of a bluriness problem as other icons when scaled down).

**Comment by fvsch** [comment]
Alternative design idea, if we want to have a single button to save some space:

1. Use a 3-state button, but make its state visible by changing the icon rather than just background and fill colors:
    - Default state: exact icon to be figured out, button has normal colors
    - Forced light mode: sun icon, button has highlighted colors (blue fill/background)
    - Forced dark mode: moon icon, button has highlighted colors (blue fill/background)

2. The implementation could stay the same: click to cycle between `["default", "light", "dark"]`. Later on, if user feedback warrants it, implement a dropdown menu to make those choices more explicit for users:
    - Normal click: show dropdown menu (see Honza’s mockup)
    - Shift + click: cycle between states, without showing the dropdown menu

**Comment by SebastianZ** [comment]
To clarify the current status:

The UI got changed in [bug 1678949](https://bugzilla.mozilla.org/show_bug.cgi?id=1678949) to use two buttons reflecting three states:

* Default state: no button clicked => feature disabled
* Forced light mode: sun icon button
* Forced dark mode: moon icon button

This feature was enabled for all channels in [bug 1692272](https://bugzilla.mozilla.org/show_bug.cgi?id=1692272) and will ship with Firefox 87.

Here's a screenshot of how it looks like now:
![Color scheme simulation buttons](https://user-images.githubusercontent.com/958943/108261222-9c205e00-7163-11eb-9254-95ab7c5b0db6.png)

Sebastian

**Comment by SebastianZ** [comment]
Personally, I'd rather aim to create a button that opens a panel for all types of simulations (currently color scheme simulation and print media simulation).

That is more consistent with the other features like the `:hov` and `.cls` buttons. And it also allows to add more media feature simulations in the future like for [`prefers-reduced-motion`](https://bugzilla.mozilla.org/show_bug.cgi?id=1477920) or [`prefers-contrast`](https://bugzilla.mozilla.org/show_bug.cgi?id=1651400).

Sebastian

**Comment by hamishwillee** [comment]
FWIW the current FF87 UI as per https://github.com/firefox-devtools/ux/issues/129#issuecomment-780819661 is confusing. 

It is fine if I toggle between the two buttons, but if toggle the dark button in dark mode (say) nothing happens and it felt like it was broken.

Perhaps hide the dark button in dark mode (and light in light mode) so that the transition is between light mode on and off (in dark mode), or dark mode on and off (in light mode). If there is a state where you might not get either mode from the OS and the browser has no default it _might_ make sense to display both buttons - but I don't know if that is possible.

**Comment by fvsch** [comment]
> An enhancement to the shipping two-button UI could be a button that opens a panel for all types of simulations (currently color scheme simulation and print media simulation). — @SebastianZ

Yep, this is considered in https://bugzilla.mozilla.org/show_bug.cgi?id=1692434

> It is fine if I toggle between the two buttons, but if toggle the dark button in dark mode (say) nothing happens and it felt like it was broken. — @hamishwillee 

I’m not sure this is something that can be fixed by changing the UI. This emulation UI needs to be able to force each one of those three values for `prefers-color-scheme`:

- `no-preference`
- `dark`
- `light`

The emulation feature cannot control how the site will react to each of those values. For instance, if a site uses a dark theme by default, and your OS is using a dark theme, the site may show up with a dark theme. When you enable the `prefers-color-scheme: no-preference` emulation in DevTools, then the site may still show a dark theme, so as you said: “nothing happens”. Something *did* happen, but the site treated it the same way as before.

There are more tricky implementations where the theme preference was saved in localStorage by the site, and it's now using that localStorage value as its source of truth instead of `prefers-color-scheme` media queries. Then changing the `prefers-color-scheme` value in DevTools will never change the page’s styles. Ironically, the DevTools UI has the same problem, since it stores the light/dark theme preference as a Firefox preference (see `devtools.theme` in about:config), and doesn't inherit the light/dark value from the Firefox theme or the OS (it used to, but at the time it proved too buggy way, so that was removed).

See also this post for some design considerations on dark/light mode toggles:
https://kilianvalkhof.com/2020/design/your-dark-mode-toggle-is-broken/

**Comment by hamishwillee** [comment]
So is "no preference" equivalent to "not light" and "not dark", or "whichever or light or dark the operating system gives you"?

**Comment by fvsch** [comment]
> So is "no preference" equivalent to "not light" and "not dark", or "whichever or light or dark the operating system gives you"?

Depends on if by "no preference" you mean "not forcing a light/dark value in DevTools" or "the page will match the `(prefers-color-scheme: no-preference)` media query". Turns out they’re pretty different things, and a source of confusion for many I’m sure.

Here’s a quick page for testing:
https://codepen.io/fvsch/pen/qBqxaVy

- If I activate the “Sun” button and reload the page, it matches `light`
- If I activate the “Moon” button and reload the page, it matches `dark`
- If I activate neither and reload the page, it matches `light` when my OS setting is on light, `dark` when my OS setting is on dark (Firefox Nightly on macOS 11.2)

It looks like Firefox does not support matching `no-preference` at this time? There is a `layout.css.prefers-color-scheme-no-preference.enabled=false` preference in about:config, but setting it to true does not seem to change things.

@nchevobbe Can you help clarify what Firefox + DevTools can do at the moment?

I wonder if ultimately DevTools should show 4 options in a menu:

- Default (light*)
- no-preference
- light
- dark

* In parenthesis, show the current value inherited from the OS/Firefox (`light`, `dark` or `no-preference`)

**Comment by hamishwillee** [comment]
So why do we expect `no-preference` to be valid when, unlike for many other queries, [it is not defined in the spec](https://drafts.csswg.org/mediaqueries-5/#prefers-color-scheme)?
In the cases where it is defined it means "web default" which in this case would be "light". 

Something else to think about though; the spec indicates that in future there might be other preferences than light or dark, which implies either a button to cycle or a set of all possibilities would be better than this odd 2 button thingy.

Anyway, three are three states here - dark set, light set, no styling set.  We want the user to be able to cycle through those three states without having to change the system mode. If they can do that with one button then awesome. Perhaps the missing part is telling them why nothing appears to be happening in particular settings.

**Comment by fvsch** [comment]
@hamishwillee I wasn’t aware of that spec change.

> FWIW then why are we trying to support it in this tool?

Define "we" and "trying" :)

I — a sometimes contributor to Firefox DevTools UX — was suggesting to add support for forcing `(prefers-color-scheme: no-preference)` to DevTools, but since that value is not in the spec anymore that's moot.

**Comment by hamishwillee** [comment]
Phew, that should make things very much easier :-).  Though as above there is this idea in the spec that more complicated colour modes like "sepia" should be supported, which would make this hard again. Thanks for your patience with me; I'm not so good at explaining things that I don't "get".

**Comment by fvsch** [comment]
No worries. Both the spec and web developers’ understanding have been a moving target, so it’s hard to get a design that’s quite right.

IMO the current design could be made clearer, e.g. by showing in some way that the “off” state means that the browser is using a default `prefers-color-scheme` value and showing what that value is. So, in a way: it could be both an inspection tool and an emulation tool.

That won’t solve all comprehension problems. Developers may still wonder why choosing “dark” or “light” doesn’t make the site dark or light (possible reasons: site doesn’t support it, or site caches a value in JavaScript memory and/or in localStorage and uses that, not reacting to further changes from the browser). I don’t think these issues can be solved without forcing sites to support all color preferences and to respond to color preference changes (which won’t happen).

But telling developers “Hey, this is the `prefers-color-scheme` value currently reported by the browser to this web page” sounds like something that is the responsibility of DevTools, and right now it’s not doing it explicitly enough (or at all, when *not* using emulation).

**Comment by SebastianZ** [comment]
Quoting the [hint in the spec.](https://drafts.csswg.org/mediaqueries-5/#prefers-color-scheme) here:

> This feature, like the other prefers-* features, previously had a no-preference value to indicate an author not expressing an active preference. However, user agents converged on expressing the "default" behavior as a light preference, and never matching no-preference.

Note that this is actually not totally correct. As you already correctly recognized, default behavior of user agents is to take the preference from the OS. So "light" is not the default.

The main confusion obviously comes from not making it clear what the default is. So I agree with @fvsch that this info should be explosed somehow.

I think the DevTools should always provide the user with all options - for now that means no emulation (use default), dark preference emulation, light preference emulation. Otherwise authors would wonder why there is only one emulation mode shown.

So for an unambiguous UI I see three options: An options button, a drop down list, or radio buttons.

An options button means a toolbar button that opens a menu that allows to choose between the options. This could be used as long as the emulation features are not grouped (see #134). That allows the button to be placed individually and at the same time have a consistent UI with the other toolbar buttons.

A drop down list is similar to the button but displays text instead of an icon (see the network throttling list as an example). That makes it obvious whether emulation is enabled and which one.

The radio buttons variant is best when the different emulation options are grouped and using some kind of panel as the radio buttons take some space.

I'll try to come up with some mockups as for how I imagine those three UI variants.

Sebastian

**Comment by SebastianZ** [comment]
Ok, here are quick mockups of the three mentioned UI variants:

Options button
-----------------
![Color scheme simulation options button](https://user-images.githubusercontent.com/958943/109715634-30e37c80-7ba4-11eb-9ef3-399a5f003b6f.png)

Drop down list
----------------
![Color scheme simulation drop down list](https://user-images.githubusercontent.com/958943/109715632-304ae600-7ba4-11eb-9e48-40e2e96b7c06.png)

Radio buttons
---------------
![Color scheme simulation radio buttons](https://user-images.githubusercontent.com/958943/109715649-350f9a00-7ba4-11eb-886c-567078569b87.png)

Please note that they are really just meant to illustrate what I meant with the three types. It still needs to be clarified
* whether the options should stay within the Rules view or be placed somewhere else.
* what icon should be used in the options button and radio buttons variant and whether it should change depending on the selected option.
* whether the icons should also be shown besides the options (probably just possible in the radio buttons solution).
* how an active emulation option is displayed.
* how grouped emulation features will look like. (That's covered in #134.)

Sebastian

**Comment by hamishwillee** [comment]
> Developers may still wonder why choosing “dark” or “light” doesn’t make the site dark or light 

True. So you might hide the tools if there is no media query of this type on the page. 

I still don't see the need for more than two states though - no emulation is never going to show you anything different, so why bother.
